### PR TITLE
AI add strategic bombing

### PR DESCRIPTION
**Pull request by ron-murhammer**
Add strategic bombing ability to AI. It mostly only uses it if it has bombers left over after attacking. Pretty basic calculations at this point and focused on revised rules currently. Future iterations will expand to properly cover v3 rules, etc.

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/860)

### CODE REVIEW

File: src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java [hunk 1]
```diff
@@ -271,6 +272,39 @@ public static void calculateBombardMoveRoutes(final PlayerID player, final List<
     }
   }
 
+  public static void calculateBombingRoutes(final PlayerID player, final List<Collection<Unit>> moveUnits,
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java]: `moveRoutes` is an out parameter, please remove it from the parameter list and return the calculated values.

File: src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java [hunk 2]
```diff
@@ -194,7 +194,8 @@ public static void calculateAmphibRoutes(final PlayerID player, final List<Colle
                   && maxUnitDistance <= minUnitDistance
                   && distanceFromUnloadTerritory < movesLeft
                   && (maxUnitDistance < minUnitDistance
-                      || (maxUnitDistance > 1 && neighborDistanceFromEnd > maxDistanceFromEnd) || (maxUnitDistance <= 1 && neighborDistanceFromEnd < maxDistanceFromEnd))) {
+                      || (maxUnitDistance > 1 && neighborDistanceFromEnd > maxDistanceFromEnd)
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java]: This if statement is complex. This will be hard to maintain, and complex boolean conditions have a higher correlation with defective code. Is there anything that can be done? 
The series of 'if' statements suggest a decomposition into polymorphic objects refactor could simplify this.

File: src/games/strategy/triplea/ai/proAI/data/ProTerritory.java [hunk 3]
```diff
@@ -432,4 +435,12 @@ public void setMaxScrambleUnits(final List<Unit> maxScrambleUnits) {
   public List<Unit> getMaxScrambleUnits() {
     return maxScrambleUnits;
   }
+
+  public List<Unit> getBombers() {
+    return bombers;
+  }
+
+  public void setBombers(final List<Unit> bombers) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/data/ProTerritory.java]: - Why is bombers mutable?
- I do not see where this is used

File: src/games/strategy/triplea/ai/proAI/data/ProMyMoveOptions.java [hunk 4]
```diff
@@ -56,4 +59,8 @@ public ProMyMoveOptions(final ProMyMoveOptions myMoveOptions) {
     return transportList;
   }
 
+  public Map<Unit, Set<Territory>> getBomberMoveMap() {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/data/ProMyMoveOptions.java]: I think there could be a number of improvements made here. Some suggestions:
- Use guava `BiMap<Unit, Territory>` instead of `Map<Unit, Set<Territory>>`
- Poor abstraction, direct access to data representation, which is also poor encapsulation. Simple DataBean classes are an anti-pattern. For example, this move map is accessed in only a couple of ways. 

For example, consider the code below - most that could be moved to this class. There would be greater cohesion, since things related to the bomber move map would be in one class. As is, the bomber move map logic can be scattered anywhere throughout the code. So, the below seems like it would fit in a method similar to: `boolean canAnyTerritoriesBeBombed(Unit)`. It looks like only the one parameter would be needed.

```
+      // Check to see if any territories can be bombed
 +      final Map<Unit, Set<Territory>> bomberMoveMap = territoryManager.getAttackOptions().getBomberMoveMap();
 +      for (final Unit unit : bomberMoveMap.keySet()) {
 +        if (alreadyAttackedWithUnits.contains(unit)) {
 +          continue;
 +        }
 +        Optional<Territory> maxBombingTerritory = Optional.empty();
 +        int maxBombingScore = 3;
 +        for (final Territory t : bomberMoveMap.get(unit)) {
 +          final boolean territoryCanBeBombed = t.getUnits().someMatch(Matches.UnitCanProduceUnitsAndCanBeDamaged);
 +          if (territoryCanBeBombed && canAirSafelyLandAfterAttack(unit, t)) {
 +            final int noAABombingDefense = t.getUnits().someMatch(Matches.UnitIsAAforBombingThisUnitOnly) ? 0 : 1;
 +            int maxDamage = 0;
 +            final TerritoryAttachment ta = TerritoryAttachment.get(t);
 +            if (ta != null) {
 +              maxDamage = ta.getProduction();
 +            }
 +            final int numExistingBombers = attackMap.get(t).getBombers().size();
 +            final int remainingDamagePotential = maxDamage - 3 * numExistingBombers;
 +            final int bombingScore = (1 + 9 * noAABombingDefense) * remainingDamagePotential;
 +            if (bombingScore > maxBombingScore) {
 +              maxBombingScore = bombingScore;
 +              maxBombingTerritory = Optional.of(t);
 +            }
 +          }
 +        }
```

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 5]
```diff
@@ -1630,4 +1686,17 @@ private void logAttackMoves(final List<ProTerritory> prioritizedTerritories) {
...
     }
   }
+
+  private boolean canAirSafelyLandAfterAttack(final Unit unit, final Territory t) {
+    final boolean isAdjacentToAlliedFactory =
+        Matches.territoryHasNeighborMatching(data,
+            ProMatches.territoryHasInfraFactoryAndIsAlliedLand(player, data)).match(t);
+    final int range = TripleAUnit.get(unit).getMovementLeft();
+    final int distance =
+        data.getMap().getDistance_IgnoreEndForCondition(ProData.unitTerritoryMap.get(unit), t,
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Would you mind adding better APIs, and calling those?
- `getDistance_IgnoreEndForCondition` is not lower camel cased
- `territoryCanMoveAirUnitsAndNoAA` takes a boolean parameter, a new API with no `boolean` parameter can replace the call and will simplify the client. Alternatively please replace the `boolean` parameter with an `Enum`

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 6]
```diff
@@ -1630,4 +1686,17 @@ private void logAttackMoves(final List<ProTerritory> prioritizedTerritories) {
...
+            ProMatches.territoryCanMoveAirUnitsAndNoAA(player, data, true));
+    final boolean usesMoreThanHalfOfRange = distance > range / 2;
+    return isAdjacentToAlliedFactory || !usesMoreThanHalfOfRange;
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: This return statement indicates there is a lot of unnecessary execution, and a SRP violation. We know very early if `isAdjacentToAlliedFactory` is true or not.  I recommend making `canAirSafelyLandAfterAttack` dead simple, and using methods rather than variable names to encapsulate the logic, ie:
`return (canAirSafelyLandAfterAttack() || !usesMoreThanHalfOfRange())`

Then in turn the two helper methods are more simple (KISS principle)

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 7]
```diff
@@ -695,6 +711,52 @@ private void determineUnitsToAttackWith(final List<ProTerritory> prioritizedTerr
...
+              maxBombingScore = bombingScore;
+              maxBombingTerritory = Optional.of(t);
+            }
+          }
+        }
+        if (maxBombingTerritory.isPresent()) {
+          final Territory t = maxBombingTerritory.get();
+          attackMap.get(t).getBombers().add(unit);
+          sortedUnitAttackOptions.remove(unit);
+          ProLogger.debug("Add bomber (" + unit + ") to " + t);
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: The debug statement is an indicator of manual testing. 'test once, never again'  The lack of test puts no pressure on the code design, and there are some problems:
- long method code smell
- nested for loops 
- nested if statements (within nested for loops, very high cyclomatic complexity)
- lack of abstraction, responsibilities are not delegated

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 8]
```diff
@@ -695,6 +711,52 @@ private void determineUnitsToAttackWith(final List<ProTerritory> prioritizedTerr
...
+      }
+
+      // Check to see if any territories can be bombed
+      final Map<Unit, Set<Territory>> bomberMoveMap = territoryManager.getAttackOptions().getBomberMoveMap();
+      for (final Unit unit : bomberMoveMap.keySet()) {
+        if (alreadyAttackedWithUnits.contains(unit)) {
+          continue;
+        }
+        Optional<Territory> maxBombingTerritory = Optional.empty();
+        int maxBombingScore = 3;
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Why declare this here? Looks like it is declared far away from its usage.
What does the number `3` represent? Looks like the `3` should be a named constant.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 9]
```diff
@@ -695,6 +711,52 @@ private void determineUnitsToAttackWith(final List<ProTerritory> prioritizedTerr
       Map<Unit, Set<Territory>> sortedUnitAttackOptions =
           tryToAttackTerritories(prioritizedTerritories, alreadyMovedUnits);
 
+      // Clear bombers
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Data encapsulation violation, looks like it should be a `clearBombers()` method that does this functionality. A `attackMap` 1st class object looks like a way to accomplish that.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 10]
```diff
@@ -157,11 +160,23 @@ public void doMove(final Map<Territory, ProTerritory> attackMap, final IMoveDele
...
 
-    // Calculate attack routes and perform moves
+    // Calculate bombard routes and perform moves
     moveUnits.clear();
     moveRoutes.clear();
     ProMoveUtils.calculateBombardMoveRoutes(player, moveUnits, moveRoutes, attackMap);
     ProMoveUtils.doMove(moveUnits, moveRoutes, null, moveDel);
+
+    // Calculate bombing routes and perform moves
+    moveUnits.clear();
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Yikes, this is a lot of state we are managing for other objects. Can we avoid any of this data state?

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 11]
```diff
@@ -157,11 +160,23 @@ public void doMove(final Map<Territory, ProTerritory> attackMap, final IMoveDele
...
     ProMoveUtils.doMove(moveUnits, moveRoutes, transportsToLoad, moveDel);
 
-    // Calculate attack routes and perform moves
+    // Calculate bombard routes and perform moves
     moveUnits.clear();
     moveRoutes.clear();
     ProMoveUtils.calculateBombardMoveRoutes(player, moveUnits, moveRoutes, attackMap);
     ProMoveUtils.doMove(moveUnits, moveRoutes, null, moveDel);
+
+    // Calculate bombing routes and perform moves
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Comment is redundant to the code

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 12]
```diff
@@ -157,11 +160,23 @@ public void doMove(final Map<Territory, ProTerritory> attackMap, final IMoveDele
...
+    moveUnits.clear();
+    moveRoutes.clear();
+    isBombing = true;
+    ProMoveUtils.calculateBombingRoutes(player, moveUnits, moveRoutes, attackMap);
+    ProMoveUtils.doMove(moveUnits, moveRoutes, null, moveDel);
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Please do not pass `null` as a method arg (for APIs we control)

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java [hunk 13]
```diff
@@ -157,11 +160,23 @@ public void doMove(final Map<Territory, ProTerritory> attackMap, final IMoveDele
...
+    // Calculate bombard routes and perform moves
     moveUnits.clear();
     moveRoutes.clear();
     ProMoveUtils.calculateBombardMoveRoutes(player, moveUnits, moveRoutes, attackMap);
     ProMoveUtils.doMove(moveUnits, moveRoutes, null, moveDel);
+
+    // Calculate bombing routes and perform moves
+    moveUnits.clear();
+    moveRoutes.clear();
+    isBombing = true;
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Is the following correct:
- we have a boolean variable that is set to one value on line 172, and another on line 175.
- The same variable is available via a public getter
- We have calls to two static methods, on lines 173 and 174
- `this` object is not passed as a parameter to any of those methods

This either means:
A) The setting of the variable has no effect
B) We are passing in an object to one of static methods, that has a `this` reference, which then uses data in this class, to presumably do something. Ouch :(

Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: More directly:
- this is very difficult to understand

I think in large part because of reliance on side effects, but also I suspect due to a lack of decomposition and abstraction.

**Comment by DanVanAtta** [comment]
"make it right, make it clean, make it fast" 


I hacked a WWII map to have 100 bombers starting instead of 1. Letting the AI play against itself, it crashes on the Japanese turn with:
```
Exception: Index: 0, Size: 0
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at games.strategy.engine.data.Route.<init>(Route.java:40)
	at games.strategy.triplea.delegate.Matches.UnitHasEnoughMovementForRoute(Matches.java:1732)
	at games.strategy.triplea.ai.proAI.ProTechAI.findPlaneAttackersThatCanLand(ProTechAI.java:504)
	at games.strategy.triplea.ai.proAI.ProTechAI.getStrengthOfPotentialAttackers(ProTechAI.java:180)
	at games.strategy.triplea.ai.proAI.ProTechAI.tech(ProTechAI.java:47)
	at games.strategy.triplea.ai.proAI.ProAI.tech(ProAI.java:277)
	at games.strategy.triplea.ai.AbstractAI.start(AbstractAI.java:432)
	at games.strategy.engine.framework.ServerGame.waitForPlayerToFinishStep(ServerGame.java:528)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:422)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:274)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInNewThread(LocalLauncher.java:63)
	at games.strategy.engine.framework.startup.launcher.AbstractLauncher.lambda$launch$195(AbstractLauncher.java:42)
	at java.lang.Thread.run(Thread.java:745)
```

(and we are getting a lot of NPE in line drawer, which I assume is fixed on latest master, and this branch is older than the fix)


I do think it is relatively interesting how the AI uses the bombers, and it does use them now to bomb:
battle_combat
bombers_landing

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 11 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Removed comments

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 10 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Not without a pretty big refactor.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 13 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Yep its actually pretty bad but this is the best way I could think of to do it. The reason for this is when you move a bomber to a territory that can be bombed you get a popup (as a human) asking whether you want to bomb or attack. The same thing happens for the AI so as it calling the delegate to move bombers, a call back from the engine occurs where the AI needs to return bombing=true.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 12 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Updated

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 9 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: So ProTerritoryManager is essentially that object. But its already become too bloated and needs broken up.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 8 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Tracking the max bombing score across the loop so needs to be before the loop. Added constant for the value.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 7 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: This is AI logging so you can see after simulating games what the AI is doing. Agree though the method/class as a whole needs broken up. That is beyond this change though.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 5 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: I'd prefer to keep AI PRs focused just on AI code and not touch engine code. That should be a separate PR. In this case that boolean is passed much further down the chain. It will also make this method name even longer. I think its better to leave it for now.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 6 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Given this method is pretty simple and only a few lines and those helper methods would only be called from this method I don't think its worth breaking it up more. You could add an if to check them in series rather than parallel but I think that actually makes it harder to read/maintain.

File: src/games/strategy/triplea/ai/proAI/data/ProMyMoveOptions.java (hunk 4 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/data/ProMyMoveOptions.java]: So the problem is if you move all the code that is touching TerritoryManager variables into it you'll have 1 class for most of the AI with 10000+ lines. TerritoryManager is representing the AIs view of the map. It tracks which units can move where and where the AI chooses to move units. In this case, it actually does makes sense to have what is primarily a data object to track that state.

File: src/games/strategy/triplea/ai/proAI/data/ProTerritory.java (hunk 3 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/data/ProTerritory.java]: AI adds bombers one by one to the list. This tracks which bombers are going to bomb which territories.

File: src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java (hunk 2 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java]: Agree, though just auto-formatting here.

File: src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java (hunk 1 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java]: So is moveUnits. It changes 2 variables so would need to create a new class in order to return them both.

**Comment by ron-murhammer** [comment]
@DanVanAtta Interesting, as that code has nothing to do with these changes and is actually the old StrongAI tech code. That issue probably exists in the current master. I added a check to avoid it though.

The AI looks to handle the somewhat extreme test scenario pretty well by spreading its bombers around to both attack and bomb. It should probably use more of them to ensure 1 round battles but not a big issue. You could probably argue that it might not pick the best noncombat move landing spot for most of them but that wasn't changed here.

I believe I addressed all the points above. Some I made some changes for the rest I would either consider out of scope or debatable.

File: src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java (hunk 1 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProMoveUtils.java]: Move this entire method into that class as well? Then it'll have some meat to it

File: src/games/strategy/triplea/ai/proAI/data/ProTerritory.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/data/ProTerritory.java]: Ah, so it does "getBombers().add()"

Would it not be cleaner to then have a `addBomber(Bomber bomber)` method?
Meanwhile, this set method is not used, and can be removed either way

File: src/games/strategy/triplea/ai/proAI/data/ProMyMoveOptions.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/data/ProMyMoveOptions.java]: I'm not necessarily recommending to move any code other than what is listed, just about 30 lines. Part of it is decomposing the state into some sort of abstraction, so it does not all have to be so low level and the state represented mostly as primitives.

Ideally things would be broken up into a number of modules. So it will be weird to have just a piece of this large class broken up when most of it is not. But that is on the right path, and makes the decomposed stuff simpler, which is easier to understand, less buggy (and testable)

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 5 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: That method is only called by AI code.The only reason to touch it, or any variation of it, is if you were modifying AI code. Adding a second API that is simply removes the one boolean param, and calls the original method with a value of true, will always be exactly 3 lines of code. Here it is for you:

```
 public static Match<Territory> territoryCanMoveAirUnitsAndNoAA(final PlayerID player, final GameData data) {
  return territoryCanMoveAirUnitsAndNoAA(player, data);
}
```

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 6 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: I disagree. Reading this method, it akes time to udnerstand every line up to 1699. Then you get to learn that you wasted time reading figuring out 3 lines of code, and wasted computation as it had no effect. It'd be better to have multiple return points than to have unnecessary execution. The simpler helper methods would make it all simple, which I think is more ideal

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 7 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Why is that beyond this change? It seems like this change is not quite complete, it has not been revised to be clean, it has not been made modular. Adding in future changes are likely going to be really painful. For example, bomber range is extended when bombers can land on carriers. Adding that in, as is, I don't think would be so pleasant. But also, given the complexity, the chance for bugs is greater than if things were simpler and unit tested.

File: src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java (hunk 9 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/ProCombatMoveAI.java]: Fair enough on this one.

**Comment by DanVanAtta** [comment]
To be frank @ron-murhammer , I think this is a very procedural addition and is not very clean. I'm in conflict as this is an ongoing battle we have. While I see the value in what has been added, at the same time it is going down the same path that made this codebase overall a large mess (still being very frank).

I think we've come a long way to fixing it, but at some point you need to start refactoring your work and adding clean code. Only then can we really begin to effectively pay down the technical debt. There should be mostly new classes for the code added in this PR. Open-close principle and SRP should be applied.

I know you have an aversion to testing, so if at least simply break this up into multiple classes, without test, I am willing to accept it. No way should this much logic be added without at least one new class. IE: if you aren't willing to test, then at least have testable code..

**Comment by DanVanAtta** [comment]
Regarding behavior - it's interesting AI uses the bombers for defensive fodder almost. Yeah, stacking the battles, particularly in dice, would be of benefit, but using an appropriate amount is good. It also looks to wisely use 2 bombers on london, and 0 in karelia. 

We should probably check out how the AI handles the dual use planes that can bomb or attack, and land on carriers (as another test case)

**Comment by DanVanAtta** [comment]
Rephrasing, I think we are still on a relatively procedural path here that has led to things being messy. To what extent more is it reasonable to refactor this to clean up the things we are adding? 

I re-reviewed the 4 commits added, it addresses a number of issues pointed out, and I agree the others would require substantial change..
